src/backend/aspen/api/views/auspice.py

```diff
@@ -0,0 +1,204 @@
+    except sa.exc.NoResultFound:  # type: ignore
+        return False, None
+    return True, phylo_tree
+
+
+def _sample_filter(sample: Sample, can_see_pi_group_ids: Set[int], system_admin: bool):
+    if system_admin:
+        return True
+    return sample.submitting_group_id in can_see_pi_group_ids
+
+
+async def _get_and_filter_phylo_tree(
+    db: AsyncSession, user: User, phylo_tree_id: int
+) -> dict:
+    authorized, phylo_tree_result = await _verify_phylo_tree_access(
+        db, user, phylo_tree_id, load_samples=True
+    )
+    if not authorized or not phylo_tree_result:
+        raise ex.BadRequestException("No phylo run found for auspice request")
+    phylo_tree: PhyloTree = phylo_tree_result
+
+    can_see_pi_group_ids: Set[int] = {user.group_id}
+    can_see_pi_group_ids.update(
+        {
+            can_see.owner_group_id
+            for can_see in user.group.can_see
+            if can_see.data_type == DataType.PRIVATE_IDENTIFIERS
+        }
+    )
+
+    identifier_map: Mapping[str, str] = {
+        sample.public_identifier.replace("hCoV-19/", ""): sample.private_identifier
+        for sample in phylo_tree.constituent_samples
+        if _sample_filter(sample, can_see_pi_group_ids, user.system_admin)
+    }
+
+    s3 = boto3.resource(
+        "s3",
+        endpoint_url=os.getenv("BOTO_ENDPOINT_URL") or None,
+        config=boto3.session.Config(signature_version="s3v4"),
+    )
+
+    data = (
+        s3.Bucket(phylo_tree.s3_bucket).Object(phylo_tree.s3_key).get()["Body"].read()
+    )
+    json_data = json.loads(data)
+
+    json_data = _rename_nodes_on_tree([json_data["tree"]], identifier_map, "GISAID_ID")
+
+    return json_data
+
+
+# TODO: Convert to POST request
+@router.post("/generate")
+async def generate_auspice_string(
+    request: Request,
+    db: AsyncSession = Depends(get_db),
+    settings: Settings = Depends(get_settings),
+    user: User = Depends(get_auth_user),
+):
+    request_body = await request.json()
+    validated_body = GenerateAuspiceMagicLinkRequest.parse_obj(request_body)
+    phylo_tree_id = request_body["tree_id"]
+    authorized_tree_access, _phylo_tree = await _verify_phylo_tree_access(
+        db, user, phylo_tree_id
+    )
+    if not authorized_tree_access:
+        raise ex.BadRequestException(
+            "No phylo run found for user to generate auspice request"
+        )
+
+    expiry_time = datetime.now(timezone.utc) + timedelta(hours=1)
+    payload = {
+        "tree_id": phylo_tree_id,
+        "user_id": user.id,
+        "expiry": expiry_time.isoformat(),
+    }
+
+    # encode before hashing
+    bytes_payload = json.dumps(payload).encode("utf8")
+    message = urlsafe_b64encode(bytes_payload)
+    # our mac key is stored as a b64encoded string
+    mac_key = urlsafe_b64decode(settings.AUSPICE_MAC_KEY)
+    digest_maker = hmac.new(mac_key, message, hashlib.sha3_512)
+    mac_tag = digest_maker.hexdigest()
```

Comment:
We don't have to do this right now, but eventually we probably want to add these sign/authenticate code blocks to a shared library.